File: agent-driven-architecture-main/examples/kotlin/block/artifact/src/main/kotlin/adr/blocks/artifact/Slice.kt
// ── blocks/artifact/slice — the work product as a FOLDED SLICE (G16) ───────
// Because the content IS State, it re-folds, it diffs by value, and it is
// crash-recoverable for free. The regression the old shape could not catch — a
// reducer change that truncates a line while leaving everything else identical — now
// fails the golden STATE assertion.

package adr.blocks.artifact

import adr.contract.ArtifactResult.ArtifactLine
import adr.spine.pure.Authority
import adr.spine.pure.Timestamp

// `ArtifactLine` is declared on this block's own sealed transport root, in
// blocks/artifact/Contract.kt: `ArtifactEffect.DeliverArtifact` carries a list of them,
// and Kotlin's sealed rule authors that file inside `:spine` (ADR-001 §3). C2 admits
// the import by name prefix — `ArtifactResult.ArtifactLine` starts with `Artifact`.

sealed class SealStatus {
    data object Draft : SealStatus()

    /** Reversible: a request is just a request. Records WHO ASKED. */
    data class Sealing(val requestedBy: Authority) : SealStatus()

    data class Sealed(val at: Timestamp, val by: Authority) : SealStatus()
}

data class ArtifactSlice(
    val lines: List<ArtifactLine> = emptyList(),
    val seal: SealStatus = SealStatus.Draft,
) {
    // No companion: a companion member has no instance, the same defect as a top-level
    // function. The EMPTY slice is what the primary constructor builds when told
    // nothing — `ArtifactSlice()` — so the shape carries its own starting value.
    fun withLine(line: ArtifactLine): ArtifactSlice = copy(lines = lines + line)

    fun withSeal(next: SealStatus): ArtifactSlice = copy(seal = next)
}

File: agent-driven-architecture-main/examples/typescript/src/blocks/artifact/slice.ts
// ── blocks/artifact/slice — THE ARTIFACT IS A FOLDED SLICE (G16) ───────────
// One line per fold arm. Because the content is State it re-folds, it diffs by
// value, and it is crash-recoverable for free. The regression the old shape
// could not catch — a reducer change that truncates a line while State stays
// byte-identical — is now impossible, because the line IS the State.
//
// `SealStatus` follows `TicketStatus`'s pattern: the parent declares
// `requestedBy`, so the boundary gate reads "is a seal pending, and who asked?"
// off any variant without a match, and a new variant must answer it by
// construction.

import type { Actor, Authority } from "@adr/spine/pure/actor";
import type { Timestamp } from "@adr/spine/pure/ids";

export interface ArtifactLine {
  readonly at: Timestamp;
  readonly by: Actor;
  readonly text: string;
}

export type SealStatusKind = "Draft" | "Sealing" | "Sealed";

export interface SealStatusBase {
  readonly kind: SealStatusKind;
  readonly requestedBy: Authority | null;
}

export interface Draft extends SealStatusBase {
  readonly kind: "Draft";
  readonly requestedBy: null;
}

export interface Sealing extends SealStatusBase {
  readonly kind: "Sealing";
  readonly requestedBy: Authority;
}

export interface Sealed extends SealStatusBase {
  readonly kind: "Sealed";
  readonly requestedBy: null;
  readonly at: Timestamp;
  readonly by: Authority;
}

export type SealStatus = Draft | Sealing | Sealed;

export const draft: Draft = { kind: "Draft", requestedBy: null };

export function sealing(requestedBy: Authority): Sealing {
  return { kind: "Sealing", requestedBy };
}

export function sealed(at: Timestamp, by: Authority): Sealed {
  return { kind: "Sealed", requestedBy: null, at, by };
}

export interface ArtifactSlice {
  readonly lines: readonly ArtifactLine[];
  readonly seal: SealStatus;
}

export const emptyArtifactSlice: ArtifactSlice = { lines: [], seal: draft };

export function withLine(slice: ArtifactSlice, line: ArtifactLine): ArtifactSlice {
  return { ...slice, lines: [...slice.lines, line] };
}

export function withSeal(slice: ArtifactSlice, seal: SealStatus): ArtifactSlice {
  return { ...slice, seal };
}
